src/main/public/js/postcode-lookup.js

```diff
@@ -0,0 +1,158 @@
+        })
+    })
+
+})
+
+function isNorthernIrelandPostcode (postcode) {
+  return postcode.startsWith('BT')
+}
+
+function showAddressEntry (postcodeLookupWidget) {
+  postcodeLookupWidget.querySelector('.address').classList.remove('js-hidden')
+}
+
+function clearPostcodeDropdown (postcodeLookupWidget) {
+  postcodeLookupWidget.querySelector('select').innerHTML = ''
+}
+
+function hidePostcodeDropdown (postcodeLookupWidget) {
+  postcodeLookupWidget.querySelector('.postcode-address-picker').classList.add('hidden')
+}
+
+function showPostcodeDropdown (postcodeLookupWidget) {
+  postcodeLookupWidget.querySelector('.postcode-address-picker').classList.remove('hidden')
+}
+
+function clearAddressFields (postcodeLookupWidget) {
+  var addressSelector = postcodeLookupWidget.querySelector('.address')
+  addressSelector.querySelector('.address-line1').value = ''
+  addressSelector.querySelector('.address-line2').value = ''
+  addressSelector.querySelector('.address-town-or-city').value = ''
+  addressSelector.querySelector('.postcode').value = ''
+}
+
+function isAnyAddressFieldPopulated(postcodeLookupWidget) {
+  var addressSelector = postcodeLookupWidget.querySelector('.address')
+
+  return addressSelector.querySelector('.address-line1').value !== '' ||
+    addressSelector.querySelector('.address-line2').value !== '' ||
+    addressSelector.querySelector('.address-town-or-city').value !== '' ||
+    addressSelector.querySelector('.postcode').value !== ''
+}
+
+function showAddressError(isNorthernIrelandPostcode, postcodeLookupWidget) {
+  var northernIrelandErrorMessage = postcodeLookupWidget.querySelector('.postcode-search-error-ni')
+  var genericErrorMessage = postcodeLookupWidget.querySelector('.postcode-search-error')
+
+  if (isNorthernIrelandPostcode) {
+    northernIrelandErrorMessage.classList.remove('hidden')
+    genericErrorMessage.classList.add('hidden')
+  } else {
+    northernIrelandErrorMessage.classList.add('hidden')
+    genericErrorMessage.classList.remove('hidden')
+  }
+
+  postcodeLookupWidget.querySelector('.postcode-search-container').classList.add('form-group-error')
+}
+
+function lookupPostcode (postcode, postcodeLookupWidget) {
+  var xhr = new XMLHttpRequest()
+  xhr.open('GET', '/postcode-lookup?postcode=' + encodeURIComponent(postcode))
+  xhr.onload = function () {
+    if (xhr.status !== 200) {
+      showAddressError(false, postcodeLookupWidget)
+      showAddressEntry(postcodeLookupWidget)
+      return
+    }
+
+    var postcodeResponse = JSON.parse(xhr.responseText)
+
+    if (!postcodeResponse.valid) {
+      var ni = isNorthernIrelandPostcode(postcode)
+      showAddressError(ni, postcodeLookupWidget)
+      showAddressEntry(postcodeLookupWidget)
+      return
+    }
+
+    var postcodeSelectDropdown = postcodeLookupWidget.querySelector('select')
+
+    var nonSelectableOption = document.createElement("option")
+    nonSelectableOption.text = postcodeResponse.addresses.length + ' addresses found'
```

Comment:
how are we going to handle translating this?

Reply:
Well spotted @robertparkinson. I suppose this has to be hidden content on the page, translated server-side?

Reply:
Either a hidden span you hide and show, or just some hidden element which acts a placeholder for text which you extract with client-side JS.